src/sentry/runner/commands/backup.py

```diff
@@ -18,7 +18,14 @@ def import_(src):
 
     for obj in serializers.deserialize("json", src, stream=True, use_natural_keys=True):
         if obj.object._meta.app_label not in EXCLUDED_APPS:
-            obj.save()
+            try:
+                obj.save()
+            # There are two integrity error exceptions we handle here.
+            # 1. Foreign key check failure
+            # 2. Duplicate key violation
```

Comment:
Isn't this one a race condition? Isn't that what we said is going on?

Reply:
Duplicate key violation is a race condition in prod, and it's handled as `except IntegrityError` in sections of the codebase. Here, it shouldn't be a race condition, but an issue with the db that should be manually fixed

Reply:
So yes and no 😁

Reply:
Plus technically, users should be trying to restore on an empty db anyways. The reason this appears is because they don't

Reply:
> The reason this appears is because they don't

If that's the case then why are we forging ahead with the restore? Don't we _want_ to halt and tell them to do something differently?

Reply:
Is that something we want to enforce? AFAIK we **strongly recommend** it [here](https://develop.sentry.dev/self-hosted/backup/) but don't enforce it. Is it reasonable to restore as much as we can and display a warning to the user?

Reply:
We definitely want to ensure that the user is warned about doing something that isn't recommended whether we decide to halt the restore or not. Technically, data is already restored leading up to when the exception is thrown. I guess the question is, should we restore some of the data, or try restoring as much as we can? I'm leaning towards restoring as much as we can but can go either way

Reply:
Decision from call is that we want to upgrade our strong recommendation to a harder requirement. Possible minor improvements to error handling would be to a) add some error messaging "This usually means you are importing into a non-clean db." and b) wrap the whole import in a transaction so that it gets rolled back on error.